pandas/core/dtypes/cast.py

```diff
@@ -275,12 +276,15 @@ def maybe_cast_result(
     dtype = maybe_cast_result_dtype(dtype, how)
 
     if not is_scalar(result):
-        if is_extension_array_dtype(dtype) and dtype.kind != "M":
-            # The result may be of any type, cast back to original
-            # type if it's compatible.
-            if len(result) and isinstance(result[0], dtype.type):
```

Comment:
This check also blocks things like the conversion of float array to nullable integer which was a problem here as well: https://github.com/pandas-dev/pandas/pull/32914